src/cunumeric/stat/histogram.cuh

```diff
@@ -0,0 +1,148 @@
+/* Copyright 2023 NVIDIA Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#pragma once
+
+#include <thrust/binary_search.h>
+#include <thrust/device_vector.h>
+#include <thrust/fill.h>
+#include <thrust/iterator/constant_iterator.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/sort.h>
+
+#include <thrust/reduce.h>
+#include <thrust/scatter.h>
+
+#include <thrust/execution_policy.h>
+#include <thrust/system/omp/execution_policy.h>
+#include <cuda_runtime.h>
+#include <thrust/version.h>
+
+#include <cub/cub.cuh>
+
+#ifdef _DEBUG
+#include <cassert>
+#include <iostream>
+#endif
+
+#include "cunumeric/utilities/thrust_util.h"
+
+#include "cunumeric/stat/histogram_gen.h"
+
+namespace cunumeric {
+namespace detail {
+template <typename element_t>
+decltype(auto) get_raw_ptr(Buffer<element_t>& v)
+{
+  return v.ptr(0);
+}
+
+// rudimentary device allocator:
+//
+template <typename elem_t, typename exe_policy_t>
+struct allocator_t<
+  elem_t,
+  exe_policy_t,
+  std::enable_if_t<!std::is_same_v<exe_policy_t, thrust::detail::host_t> &&
+                   !std::is_same_v<exe_policy_t, thrust::system::omp::detail::par_t>>> {
```

Comment:
How? I tried that. The only way to be able to reuse the same `histogram_weights()` by 3 different ecosystems (GPU/CPU/OMP) is to specialize the different moving parts (details) that are used by the generic algorithm (allocators, segemented sums, etc.)

Reply:
I'm suggesting that you only have a single templated definition, removing the `enable_if_t`. I believe the code between the three cases (CPU, OMP, GPU) can be made uniform, so no partial instantiation is necessary.

Reply:
The implementation code (`histogram_impl.h`) must see uniform class names for either the GPU or CPU/OMP branch. That implies it must see only one generic type (say, for `segmented_sum_t`). 

But because `segmented_sum_t` has completely different implementations between GPU and CPU, it must branch-out those implementations somehow, and in the same time keep its class name (type): `segmented_sum_t`. Well, the only way to branch out a generic type is via specialization (in this case partial).

Enable_if_t is just a convenient way to say: have one specialization for any policy running on device (of which can be many, not just one) and another for policies running on `host_t` or using `omp_t`. IOW, `enable_if_t` is just a convenient shortcut (making the code more succinct, hence clearer; incidentally, there are many more such instances of using `enable_if_t` in the code).

I don't see how specialization can be avoided w/o providing separate `histogram_impl_gpu.h`, `histogram_impl_cpu.h`, `histogram_impl_omp.h` (which is code replication). Can you suggest a way?

Reply:
> But because segmented_sum_t has completely different implementations between GPU and CPU

My comment above is referring to `allocator_t`, not `segmented_sum_t`. The `allocator_t` class is practically the same between CPU and GPU code, so I suggest reimplementing it as follows and moving it to the shared `histogram_impl.h`

```
template <typename elem_t, typename exe_policy_t>
struct allocator_t {
  allocator_t(void) {}

  allocator_t(elem_t, exe_policy_t) {}  // tag-dispatch for CTAD

  elem_t* operator()(exe_policy_t exe_pol, size_t size)
  {
    d_buffer_     = create_buffer<elem_t>(size);
    elem_t* d_ptr = get_raw_ptr(d_buffer_);

    return d_ptr;
  }

  elem_t* operator()(exe_policy_t exe_pol, size_t size, elem_t init)
  {
    d_buffer_     = create_buffer<elem_t>(size);
    elem_t* d_ptr = get_raw_ptr(d_buffer_);

    thrust::fill_n(exe_pol, d_ptr, size, init);

    return d_ptr;
  }

 private:
  Buffer<elem_t> d_buffer_;
};
```

Reply:
I see. Okay, done.